### ISSUE TITLE
test-remote: end-to-end alert test failed: Failed to get non-zero alerts metric value after 300s

### ISSUE THREAD

**Issue by jgehrcke** [opstrace/opstrace#1130]
```
[2021-07-28T01:03:17Z]   21 passing (9m)
[2021-07-28T01:03:17Z]   4 pending
[2021-07-28T01:03:17Z]   1 failing
[2021-07-28T01:03:17Z]
[2021-07-28T01:03:17Z]   1) End-to-end alert tests
[2021-07-28T01:03:17Z]        End-to-end alerts for default and system tenants:
[2021-07-28T01:03:17Z]      Error: Failed to get non-zero alerts metric value after 300s. Are alerts successfully reaching the e2ealerting pod?
[2021-07-28T01:03:17Z]       at waitForE2EAlertFiring (test_alerts.ts:398:13)
[2021-07-28T01:03:17Z]       at Context.<anonymous> (test_alerts.ts:458:5)
[2021-07-28T01:03:17Z] 
[2021-07-28T01:03:17Z]
```
https://buildkite.com/opstrace/prs/builds/5222#5e7d212e-6f7f-4b34-bdfa-db85f9845df3/2980-4923

**Comment by sreis** [comment]
```
[2021-07-29T09:25:48Z]   21 passing (9m)
[2021-07-29T09:25:48Z]   4 pending
[2021-07-29T09:25:48Z]   1 failing
[2021-07-29T09:25:48Z]
[2021-07-29T09:25:48Z]   1) End-to-end alert tests
[2021-07-29T09:25:48Z]        End-to-end alerts for default and system tenants:
[2021-07-29T09:25:48Z]      Error: Failed to get non-zero alerts metric value after 300s. Are alerts successfully reaching the e2ealerting pod?
[2021-07-29T09:25:48Z]       at waitForE2EAlertFiring (test_alerts.ts:398:13)
[2021-07-29T09:25:48Z]       at Context.<anonymous> (test_alerts.ts:453:5)
```
https://buildkite.com/opstrace/upgrade-tests/builds/1096#83b08d29-36ad-40d5-8d0a-e0c43056daa2/4006-5881

**Comment by jgehrcke** [comment]
Seen six days ago by @terrcin, reported here: https://github.com/opstrace/opstrace/issues/1102#issuecomment-887923388

**Comment by jgehrcke** [comment]
```
[2021-08-03T08:57:41Z]   21 passing (8m)
[2021-08-03T08:57:41Z]   4 pending
[2021-08-03T08:57:41Z]   1 failing
[2021-08-03T08:57:41Z]
[2021-08-03T08:57:41Z]   1) End-to-end alert tests
[2021-08-03T08:57:41Z]        End-to-end alerts for default and system tenants:
[2021-08-03T08:57:41Z]      Error: Failed to get non-zero alerts metric value after 300s. Are alerts successfully reaching the e2ealerting pod?
[2021-08-03T08:57:41Z]       at waitForE2EAlertFiring (test_alerts.ts:398:13)
[2021-08-03T08:57:41Z]       at Context.<anonymous> (test_alerts.ts:453:5)
[2021-08-03T08:57:41Z] 
[2021-08-03T08:57:41Z]
```
https://buildkite.com/opstrace/prs/builds/5309#402cb854-60f3-4b9f-987c-fa8fa4b15684/4490-6358

**Comment by nickbp** [comment]
https://buildkite.com/opstrace/upgrade-tests/builds/1149#3c98dcb9-c809-427a-96a8-55acdd351fce/4002-5878

**Comment by triclambert** [comment]
https://buildkite.com/opstrace/prs/builds/5432#9756a894-18d4-43db-9a3e-a09d61eaa7cf/3065-4939

**Comment by jgehrcke** [comment]
https://buildkite.com/opstrace/prs/builds/5478#6aa0eb52-386c-4bed-a132-9e5f6ad353b7/3988-5932
```
[2021-08-11T09:12:22Z]   20 passing (14m)
[2021-08-11T09:12:22Z]   4 pending
[2021-08-11T09:12:22Z]   2 failing
[2021-08-11T09:12:22Z]
[2021-08-11T09:12:22Z]   1) End-to-end alert tests
[2021-08-11T09:12:22Z]        End-to-end alerts for default and system tenants:
[2021-08-11T09:12:22Z]      Error: Failed to get non-zero alerts metric value after 300s. Are alerts successfully reaching the e2ealerting pod?
[2021-08-11T09:12:22Z]       at waitForE2EAlertFiring (test_alerts.ts:371:13)
[2021-08-11T09:12:22Z]       at Context.<anonymous> (test_alerts.ts:449:5)
```

**Comment by triclambert** [comment]
```
[2021-08-11T17:45:37Z]   21 passing (10m)
[2021-08-11T17:45:37Z]   4 pending
[2021-08-11T17:45:37Z]   1 failing
[2021-08-11T17:45:37Z]
[2021-08-11T17:45:37Z]   1) End-to-end alert tests
[2021-08-11T17:45:37Z]        End-to-end alerts for default and system tenants:
[2021-08-11T17:45:37Z]      Error: Failed to get non-zero alerts metric value after 300s. Are alerts successfully reaching the e2ealerting pod?
[2021-08-11T17:45:37Z]       at waitForE2EAlertFiring (test_alerts.ts:371:13)
[2021-08-11T17:45:37Z]       at Context.<anonymous> (test_alerts.ts:444:5)
```

https://buildkite.com/opstrace/upgrade-tests/builds/1177#3e6c5a5a-7b74-4d4b-a9ff-b9728f1540eb/1632-3518

**Comment by jgehrcke** [comment]
```
[2021-08-15T13:04:59Z]   21 passing (8m)
[2021-08-15T13:04:59Z]   4 pending
[2021-08-15T13:04:59Z]   1 failing
[2021-08-15T13:04:59Z]
[2021-08-15T13:04:59Z]   1) End-to-end alert tests
[2021-08-15T13:04:59Z]        End-to-end alerts for default and system tenants:
[2021-08-15T13:04:59Z]      Error: Failed to get non-zero alerts metric value after 300s. Are alerts successfully reaching the e2ealerting pod?
[2021-08-15T13:04:59Z]       at waitForE2EAlertFiring (test_alerts.ts:371:13)
[2021-08-15T13:04:59Z]       at Context.<anonymous> (test_alerts.ts:444:5)
```
https://buildkite.com/opstrace/scheduled-main-builds/builds/2870#4a84e8e1-1272-401c-8043-ad396226c198/3577-5461

**Comment by nickbp** [comment]
https://buildkite.com/opstrace/scheduled-main-builds/builds/2862#dc8048c5-183b-4ffe-9e4c-f855a3929792/2517-4409

**Comment by jgehrcke** [comment]
```
[2021-08-16T19:52:29Z]   21 passing (8m)
[2021-08-16T19:52:29Z]   4 pending
[2021-08-16T19:52:29Z]   1 failing
[2021-08-16T19:52:29Z]
[2021-08-16T19:52:29Z]   1) End-to-end alert tests
[2021-08-16T19:52:29Z]        End-to-end alerts for default and system tenants:
[2021-08-16T19:52:29Z]      Error: Failed to get non-zero alerts metric value after 300s. Are alerts successfully reaching the e2ealerting pod?
[2021-08-16T19:52:29Z]       at waitForE2EAlertFiring (test_alerts.ts:371:13)
```
https://buildkite.com/opstrace/prs/builds/5553#6b31a951-39cf-44e1-9823-7ee01275bca0/2879-4754

**Comment by nickbp** [comment]
build PR/5553

---

failing GCP main:

test run:
> 2021-08-16T19:44:55.531Z [32minfo[39m: Setting up E2E alerts webhook for tenant=default
> 2021-08-16T19:44:55.724Z [32minfo[39m: HTTP resp to POST(https://prs-bk-5553-cfd-g.opstracegcp.com/api/v1/alerts):
>   status: 201
>   body[:500]: 
>   headers: {"date":"Mon, 16 Aug 2021 19:44:55 GMT","content-length":"0","connection":"close","vary":"Accept-Encoding","strict-transport-security":"max-age=15724800; includeSubDomains"}
>   totalTime: 0.192 s
>   dnsDone->TCPconnectDone: 0.034 s
>   connectDone->reqSent 0 s
>   reqSent->firstResponseByte: 0.122 s
> ...
> 2021-08-16T19:44:56.604Z [32minfo[39m: Setting up E2E alerts webhook for tenant=system
> 2021-08-16T19:44:56.787Z [32minfo[39m: HTTP resp to POST(https://prs-bk-5553-cfd-g.opstracegcp.com/api/v1/alerts):
>   status: 201
>   body[:500]: 
>   headers: {"date":"Mon, 16 Aug 2021 19:44:56 GMT","content-length":"0","connection":"close","vary":"Accept-Encoding","strict-transport-security":"max-age=15724800; includeSubDomains"}
>   totalTime: 0.181 s
>   dnsDone->TCPconnectDone: 0.03 s
>   connectDone->reqSent 0 s
>   reqSent->firstResponseByte: 0.12 s
> ...
> 2021-08-16T19:45:40.596Z info: Waiting for cortex E2E alerting metric for tenant=default job=testalerts-default-i08LP-job with nonzero value
> ...
> 2021-08-16T19:50:37.046Z info: Got alert count value: 0
>     1) End-to-end alerts for default and system tenants

`alertmanager-78c87dcf7c-fswdb`:
> level=info ts=2021-08-16T19:45:01.730463512Z caller=multitenant.go:897 component=MultiTenantAlertmanager msg="updating new per-tenant alertmanager" user=system

`alertmanager-78c87dcf7c-kdwf8`:
> level=info ts=2021-08-16T19:45:00.309527332Z caller=multitenant.go:897 component=MultiTenantAlertmanager msg="updating new per-tenant alertmanager" user=system

`alertmanager-78c87dcf7c-xzwrt`:
> level=info ts=2021-08-16T19:44:59.241864505Z caller=multitenant.go:897 component=MultiTenantAlertmanager msg="updating new per-tenant alertmanager" user=system
> level=error ts=2021-08-16T19:45:09.246100224Z caller=dispatch.go:354 component=MultiTenantAlertmanager user=system component=dispatcher component=dispatcher msg="Notify for alerts failed" num_alerts=1 err="e2e-alerting/webhook[0]: notify retry canceled after 2 attempts: Post \"http://e2ealerting.system-tenant.svc.cluster.local/api/v1/receiver\": context deadline exceeded"

---

compare to passing GCP upgrade:

test run:
> 2021-08-16T19:48:33.946Z [32minfo[39m: Setting up E2E alerts webhook for tenant=default
> 2021-08-16T19:48:34.091Z [32minfo[39m: HTTP resp to POST(https://prs-bk-5553-cfd-ug.opstrace.io/api/v1/alerts):
>   status: 201
>   body[:500]: 
>   headers: {"date":"Mon, 16 Aug 2021 19:48:34 GMT","content-length":"0","connection":"close","vary":"Accept-Encoding","strict-transport-security":"max-age=15724800; includeSubDomains"}
>   totalTime: 0.143 s
>   dnsDone->TCPconnectDone: 0.033 s
>   connectDone->reqSent 0 s
> ...
> 2021-08-16T19:48:35.020Z info: Setting up E2E alerts webhook for tenant=system
> 2021-08-16T19:48:35.158Z info: HTTP resp to POST(https://prs-bk-5553-cfd-ug.opstrace.io/api/v1/alerts):
>   status: 201
>   body[:500]:
>   headers: {"date":"Mon, 16 Aug 2021 19:48:35 GMT","content-length":"0","connection":"close","vary":"Accept-Encoding","strict-transport-security":"max-age=15724800; includeSubDomains"}
>   totalTime: 0.137 s
>   dnsDone->TCPconnectDone: 0.031 s
>   connectDone->reqSent 0 s
>   reqSent->firstResponseByte: 0.072 s
> ...
> 2021-08-16T19:49:43.200Z info: Waiting for cortex E2E alerting metric for tenant=default job=testalerts-default-Mi83F-job with nonzero value
> ...
> 2021-08-16T19:51:29.844Z info: Got alerts metric value for tenant default: 1

`alertmanager-754cb7bf44-5d7gg`:
> level=info ts=2021-08-16T19:48:39.559811439Z caller=multitenant.go:897 component=MultiTenantAlertmanager msg="updating new per-tenant alertmanager" user=system
> level=info ts=2021-08-16T19:48:39.562892633Z caller=multitenant.go:897 component=MultiTenantAlertmanager msg="updating new per-tenant alertmanager" user=default
> level=error ts=2021-08-16T19:48:49.563612257Z caller=dispatch.go:354 component=MultiTenantAlertmanager user=system component=dispatcher component=dispatcher msg="Notify for alerts failed" num_alerts=2 err="e2e-alerting/webhook[0]: notify retry canceled after 2 attempts: Post \"http://e2ealerting.system-tenant.svc.cluster.local/api/v1/receiver\": context deadline exceeded"

`alertmanager-754cb7bf44-84lml`:
> level=info ts=2021-08-16T19:48:35.61222751Z caller=multitenant.go:897 component=MultiTenantAlertmanager msg="updating new per-tenant alertmanager" user=default
> level=info ts=2021-08-16T19:48:35.615172161Z caller=multitenant.go:897 component=MultiTenantAlertmanager msg="updating new per-tenant alertmanager" user=system

`alertmanager-754cb7bf44-mvrdk`:
> level=info ts=2021-08-16T19:48:37.335591315Z caller=multitenant.go:897 component=MultiTenantAlertmanager msg="updating new per-tenant alertmanager" user=system
> level=info ts=2021-08-16T19:48:37.339306936Z caller=multitenant.go:897 component=MultiTenantAlertmanager msg="updating new per-tenant alertmanager" user=default

---

Things that stick out:
- The main difference is that the failing test does not log about receiving the default tenant's alertmanager config.
- Both failing and succeeding tests log an error about the `e2ealerting` service not being reachable - so the error doesn't seem to affect the test result and might just be from when the `e2ealerting` pod is still deploying.
- Meanwhile, from the test driver's perspective, the HTTP response from calling the alertmanager config endpoint is identical across the failing and succeeding tests.
- The alertmanagers each log receiving the new configs at different times from each other, with a drift of a second or two.

From this, I would guess:
- The different times that the alertmanager pods detect the changed configs might mean that they each have some kind of poll loop to detect config updates.
- If they're polling, then they probably don't receive the update RPC directly, maybe the config is written to storage by someone else and then the alertmanagers read it from there via a poll.
- The separation between submitting a config and having the alertmanagers pick it up would also explain why submitting the config gets a 201 and then somehow doesn't get applied to the alertmanagers.

Possible fixes/things to check:
- Do a GET to check that the alertmanager config exists after submitting it? I wouldn't expect this to make a difference if the problem is in an alertmanager poll loop, which would be separate from the config storage API path.
- Remove the preceding DELETE before the post? Maybe this is confusing the alertmanager polling or otherwise creating some form of race condition that leads to the alertmanager not seeing the new config.
- Whether the config failure is consistently for the default tenant or the system tenant - in the above case the default tenant's config was dropped while the system tenant config got through.

**Comment by nickbp** [comment]
Things that stick out:
- The main difference is that the failing test does not log about receiving the default tenant's alertmanager config.
- Both failing and succeeding tests log an error about the `e2ealerting` service not being reachable - so the error doesn't seem to affect the test result and might just be from when the `e2ealerting` pod is still deploying.
- Meanwhile, from the test driver's perspective, the HTTP response from calling the alertmanager config endpoint is identical across the failing and succeeding tests.
- The alertmanagers each log receiving the new configs at different times from each other, with a drift of a second or two.

From this, I would guess:
- The different times that the alertmanager pods detect the changed configs might mean that they each have some kind of poll loop to detect config updates.
- If they're polling, then they probably don't receive the update RPC directly, maybe the config is written to storage by someone else and then the alertmanagers read it from there via a poll.
- The separation between submitting a config and having the alertmanagers pick it up would also explain why submitting the config gets a 201 and then somehow doesn't get applied to the alertmanagers.

Possible fixes/things to check:
- Do a GET to check that the alertmanager config exists after submitting it? I wouldn't expect this to make a difference if the problem is in an alertmanager poll loop, which would be separate from the config storage API path.
- Remove the preceding DELETE before the post? Maybe this is confusing the alertmanager polling or otherwise creating some form of race condition that leads to the alertmanager not seeing the new config.
- Whether the config failure is consistently for the default tenant or the system tenant - in the above case the default tenant's config was dropped while the system tenant config got through.

**Comment by nickbp** [comment]
Across failed runs, listing the tenants that were NOT configured according to alertmanager logs:
- PR/5553: default
- main/2862: default
- main/2870: default
- upgrade/1177: default
- PR/5478: neither (both default and system show up in logs - maybe this was different?)
- update/1149: default
- PR/5432: default

So that at least implies that the problem is with the default tenant - this might just be because it's the one we configure first

**Comment by nickbp** [comment]
Some followup:
- The alerts config [https://github.com/cortexproject/cortex/blob/master/pkg/alertmanager/api.go#L144](write path) goes directly to storage (S3)
- The alerts config [https://github.com/cortexproject/cortex/blob/master/pkg/alertmanager/multitenant.go#L757](read path) runs separately in a sync loop

So the config handling seems to work as expected given the behavior in the logs above.

I've been scouring over the cortex code and I don't see anything that clearly explains the flake yet. I did find some [config sync logic](https://github.com/cortexproject/cortex/blob/32b558bbe89d75514e39ecbe4f19d9bbc22ede9c/pkg/alertmanager/multitenant.go#L688-L693) but it has an `info` log entry that doesn't show up in the test logs, so it doesn't look like it's actually getting executed in these tests. But regardless of that, between deleting the config from S3, writing the new config to S3, alertmanagers polling S3, and alertmanagers syncing state among themselves, I could see a race condition occurring pretty easily.

Given this, I'll try updating the test to first check if a config exists before sending a delete request, and then to wait 5 seconds after any delete request is sent to reduce the likelihood of weird sync issues that may be caused by rapid delete+set calls.

**Comment by jgehrcke** [comment]
```
[2021-08-17T00:44:08Z]   21 passing (7m)
[2021-08-17T00:44:08Z]   4 pending
[2021-08-17T00:44:08Z]   1 failing
[2021-08-17T00:44:08Z]
[2021-08-17T00:44:08Z]   1) End-to-end alert tests
[2021-08-17T00:44:08Z]        End-to-end alerts for default and system tenants:
[2021-08-17T00:44:08Z]      Error: Failed to get non-zero alerts metric value after 300s. Are alerts successfully reaching the e2ealerting pod?
[2021-08-17T00:44:08Z]       at waitForE2EAlertFiring (test_alerts.ts:371:13)
[2021-08-17T00:44:08Z]       at Context.<anonymous> (test_alerts.ts:444:5)
```
https://buildkite.com/opstrace/scheduled-main-builds/builds/2882#d147c50c-b7bb-4309-a0c8-92e933f784e1/3265-5153

**Comment by jgehrcke** [comment]
https://buildkite.com/opstrace/prs/builds/5553#6b31a951-39cf-44e1-9823-7ee01275bca0/2879-4760

**Comment by jgehrcke** [comment]
https://buildkite.com/opstrace/prs/builds/5558#3ce7cdaa-c548-430c-b94a-ab24b8f7b23e/3007-4884
```
[2021-08-17T08:44:51Z]   21 passing (7m)
[2021-08-17T08:44:51Z]   4 pending
[2021-08-17T08:44:51Z]   1 failing
[2021-08-17T08:44:51Z]
[2021-08-17T08:44:51Z]   1) End-to-end alert tests
[2021-08-17T08:44:51Z]        End-to-end alerts for default and system tenants:
[2021-08-17T08:44:51Z]      Error: Failed to get non-zero alerts metric 
```

**Comment by nickbp** [comment]
This MIGHT be fixed via #1286 but going to wait another day to be sure. We'd been getting one of these roughly once a day recently.

**Comment by nickbp** [comment]
Haven't seen any new failures so it's looking pretty likely at this point that the above fix did the trick
